classes/userGroup/Repository.php

```diff
@@ -0,0 +1,469 @@
+            ->get()
+            ->isNotEmpty();
+    }
+
+    /**
+    * return whether a context has a specific user group
+    */
+    public function contextHasGroup(int $contextId, int $userGroupId) : bool
+    {
+        $collector = Repo::userGroup()
+            ->getCollector()
+            ->filterByContextIds([$contextId]);
+
+        $userGroups = $this->getMany($collector);
+
+        return $userGroups->where('id', $userGroupId)->count();
+    }
+
+    public function assignUserToGroup(int $userId, int $userGroupId) : UserUserGroup
+    {
+        return UserUserGroup::create([
+            'userId' => $userId,
+            'userGroupId' => $userGroupId
+        ]);
+    }
+
+    public function removeUserFromGroup($userId, $userGroupId, $contextId) : bool
+    {
+        return UserUserGroup::withUserId($userId)
+            ->withUserGroupId($userGroupId)
+            ->withContextId($contextId)
+            ->delete();
+    }
+
+    public function deleteAssignmentsByUserId(int $userId, ?int $userGroupId = null) : bool
+    {
+        $query = UserUserGroup::withUserId($userId);
+
+        if ($userGroupId) {
+            $query->withUserGroupId($userGroupId);
+        }
+
+        return $query->delete();
+    }
+
+    public function deleteAssignmentsByContextId(int $contextId, ?int $userId = null) : bool
+    {
+        $userUserGroups = UserUserGroup::withContextId($contextId);
+
+        if ($userId) {
+            $userUserGroups->withUserId($userId);
+        }
+
+        return $userUserGroups->delete();
+    }
+
+    /**
+    * Get the user groups assigned to each stage.
+    *
+    * @return LazyCollection<UserGroup>
+    */
+    public function getUserGroupsByStage($contextId, $stageId, $roleId = null, $count = null) : LazyCollection
+    {
+        $userGroups = $this->getCollector()
+            ->filterByContextIds([$contextId])
+            ->filterByStageIds([$stageId]);
+
+        if ($roleId) {
+            $userGroups->filterByRoleIds([$roleId]);
+        }
+
+        $userGroups->orderBy(Collector::ORDERBY_ROLE_ID);
+
+        return $userGroups->getMany();
+    }
+
+    /**
+    * Remove a user group from a stage
+    */
+    public function removeGroupFromStage(int $contextId, int $userGroupId, int $stageId) : bool
+    {
+        return UserGroupStage::withContextId($contextId)
+            ->withUserGroupId($userGroupId)
+            ->withStageId($stageId)
+            ->delete();
+    }
+
+    /**
+     * Get all stages assigned to one user group in one context.
+     *
+     * @param int $contextId The context ID.
+     * @param int $userGroupId The user group ID
+     *
+     * @return array
+     */
+    public function getAssignedStagesByUserGroupId(int $contextId, int $userGroupId) : array
+    {
+        return UserGroupStage::withContextId($contextId)
+            ->withUserGroupId($userGroupId)
+            ->pluck('stage_id')
+            ->toArray();
```

Comment:
I'm leaning towards having these functions return `Collection` objects, and only force conversion to arrays where it's needed.

Reply:
I changed that.